src/ray/raylet/node_manager.cc

```diff
@@ -2988,6 +3015,12 @@ MemoryUsageRefreshCallback NodeManager::CreateMemoryUsageRefreshCallback() {
   };
 }
 
+void NodeManager::SetTaskFailureReason(const TaskID &task_id, std::unique_ptr<rpc::RayException> failure_reason) {
+  RAY_LOG(INFO) << "set failure reason for task " << task_id;
+  auto result = task_failure_reasons_.emplace(task_id, std::move(failure_reason));
```

Comment:
When/how should we GC these?

Reply:
Chatted offline seems we could consider writing those records to GCS when the entry expires on the raylet, @rkooo567 thoughts?

Reply:
maybe we should record it on the owner side since we query the owner for the state API?

Reply:
per offline discussion we are keeping track of the failures in task manager, so the state api can query it